skimage/util/dtype.py

```diff
@@ -134,165 +138,162 @@ def compare(x, y, kind='u'):
 
-    if kind in 'ui':
-        imin = np.iinfo(dtype).min
-        imax = np.iinfo(dtype).max
-    if kind_in in 'ui':
-        imin_in = np.iinfo(dtype_in).min
-        imax_in = np.iinfo(dtype_in).max
-
+    # float -> any
     if kind_in == 'f':
         if np.min(image) < -1.0 or np.max(image) > 1.0:
             raise ValueError("Images of type float must be between -1 and 1.")
-        if kind == 'f':
-            # floating point -> floating point
-            if itemsize_in > itemsize:
+        if kind_out == 'f':
+            # float -> float
+            if itemsize_in > itemsize_out:
                 prec_loss()
-            return image.astype(dtype)
+            return image.astype(dtype_out)
 
         # floating point -> integer
         prec_loss()
         # use float type that can represent output integer type
-        image = np.array(image, _dtype(itemsize, dtype_in,
-                                       np.float32, np.float64))
+        image = image.astype(_dtype_itemsize(itemsize_out, dtype_in,
+                                             np.float32, np.float64))
         if not uniform:
-            if kind == 'u':
-                image *= imax
+            if kind_out == 'u':
+                image *= imax_out
             else:
-                image *= imax - imin
+                image *= imax_out - imin_out
                 image -= 1.0
                 image /= 2.0
             np.rint(image, out=image)
-            np.clip(image, imin, imax, out=image)
-        elif kind == 'u':
-            image *= imax + 1
-            np.clip(image, 0, imax, out=image)
+            np.clip(image, imin_out, imax_out, out=image)
+        elif kind_out == 'u':
+            image *= imax_out + 1
+            np.clip(image, 0, imax_out, out=image)
         else:
-            image *= (imax - imin + 1.0) / 2.0
+            image *= (imax_out - imin_out + 1.0) / 2.0
             np.floor(image, out=image)
-            np.clip(image, imin, imax, out=image)
-        return image.astype(dtype)
+            np.clip(image, imin_out, imax_out, out=image)
+        return image.astype(dtype_out)
 
-    if kind == 'f':
-        # integer -> floating point
-        if itemsize_in >= itemsize:
+    # signed/unsigned int -> float
+    if kind_out == 'f':
+        if itemsize_in >= itemsize_out:
             prec_loss()
         # use float type that can exactly represent input integers
-        image = np.array(image, _dtype(itemsize_in, dtype,
-                                       np.float32, np.float64))
+        image = image.astype(_dtype_itemsize(itemsize_in, dtype_out,
+                                             np.float32, np.float64))
         if kind_in == 'u':
             image /= imax_in
             # DirectX uses this conversion also for signed ints
-            #if imin_in:
-            #    np.maximum(image, -1.0, out=image)
+            # if imin_in:
+            #     np.maximum(image, -1.0, out=image)
```

Comment:
We'll digg into this shortly with https://github.com/scikit-image/scikit-image/issues/1234 :).